llms/openai/openaillm.go

```diff
@@ -67,35 +68,74 @@ func (o *LLM) GenerateContent(ctx context.Context, messages []llms.MessageConten
 		case schema.ChatMessageTypeGeneric:
 			msg.Role = RoleUser
 		case schema.ChatMessageTypeFunction:
-			fallthrough
+			msg.Role = RoleFunction
+		case schema.ChatMessageTypeTool:
+			msg.Role = RoleTool
+			// Here we extract tool calls from the message and populate the ToolCalls field.
+
+			// parse mc.Parts (which should have one entry of type ToolCallResponse) and populate msg.Content and msg.ToolCallID
+			if len(mc.Parts) != 1 {
+				return nil, fmt.Errorf("expected exactly one part for role %v, got %v", mc.Role, len(mc.Parts))
+			}
+			switch p := mc.Parts[0].(type) {
```

Comment:
This is odd -- seems like we should statically know whether a pointer or not is returned and error out in the other case. Which one is it?

Reply:
I agree it's a bit odd -- since both of these types satisfy isPart() the part could be either of them, hence checking both. Very much open to suggestions here.

Reply:
There are several places in the code where we type-switch on parts, e.g. https://github.com/tmc/langchaingo/blob/main/llms/googleai/googleai.go#L125

and we only handle the non-pointer type

This is part of the documentation - passing plain `TextPart` etc. is expected

If we want to match both pointer and plain everywhere, we should probably add more interfaces

Reply:
I updated this switch to only handle non-pointer values.